### 실습 1. Excel 파일 불러오기

해야 할 일

pandas를 이용해 Excel 파일을 DataFrame으로 불러옵니다.

먼저 코드부터 작성하기보다 인공지능에게 현재 상황과 원하는 작업을 설명해 봅니다.

In [1]:
import pandas as pd

df = pd.read_excel("교보문고_종합_베스트셀러_상품리스트.xlsx")

df.head()

c:\dev\llm-data-analysis-course\.venv\Lib\site-packages\openpyxl\styles\stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


,순위,상품코드,판매상품 ID,상품명,정가,판매가,할인율,적립율,적립예정포인트,인물,출판사,발행(출시)일자,분야
0,1,9788936434120,S000000610612,소년이 온다,"15,000","13,500",10%,5%,750,한강,창비,20140519,소설
1,2,9788998441012,S000001632467,모순,"13,000","11,700",10%,5%,650,양귀자,쓰다,20130401,소설
2,3,9788997780624,S000216262757,결국 국민이 합니다,"22,000","19,800",10%,5%,"1,100",이재명,오마이북,20250415,정치/사회
3,4,9788936439743,S000216111714,혼모노,"18,000","16,200",10%,5%,900,성해나,창비,20250328,소설
4,5,9788937473401,S000200550189,급류,"14,000","12,600",10%,5%,700,정대건,민음사,20221222,소설


### 실행 결과

- `pandas`의 `read_excel()`을 사용하여 교보문고 베스트셀러 Excel 파일을 DataFrame으로 불러왔다.
- 불러온 데이터는 `df`에 저장하였다.
- `df.head()`를 통해 데이터의 앞 5행과 컬럼 구성을 확인하였다.
- 순위, 상품코드, 상품명, 정가, 판매가, 할인율, 인물, 출판사 등의 정보가 포함되어 있음을 확인하였다.

### 실습 2. 데이터의 크기와 컬럼 확인하기

Excel 파일을 불러왔으면 바로 전처리를 시작하지 않습니다.

먼저 어떤 데이터인지 확인합니다.

In [2]:
# 1. 전체 행과 컬럼 개수 확인
print("1. 전체 행과 컬럼 개수")
print(df.shape)

# 2. 컬럼 이름 목록 확인
print("\n2. 컬럼 이름")
print(df.columns)

# 3. 각 컬럼의 자료형 확인
print("\n3. 각 컬럼의 자료형")
print(df.dtypes)

# 4. 각 컬럼의 결측치 개수 확인
print("\n4. 각 컬럼의 결측치 개수")
print(df.isna().sum())

# 5. 앞의 5행 확인
print("\n5. 앞의 5행")
display(df.head())

1. 전체 행과 컬럼 개수
(199, 13)

2. 컬럼 이름
Index(['순위', '상품코드', '판매상품 ID', '상품명', '정가', '판매가', '할인율', '적립율', '적립예정포인트',
       '인물', '출판사', '발행(출시)일자', '분야'],
      dtype='str')

3. 각 컬럼의 자료형
순위          int64
상품코드        int64
판매상품 ID       str
상품명           str
정가            str
판매가           str
할인율           str
적립율           str
적립예정포인트       str
인물            str
출판사           str
발행(출시)일자    int64
분야            str
dtype: object

4. 각 컬럼의 결측치 개수
순위          0
상품코드        0
판매상품 ID     0
상품명         0
정가          0
판매가         0
할인율         0
적립율         0
적립예정포인트     0
인물          0
출판사         0
발행(출시)일자    0
분야          0
dtype: int64

5. 앞의 5행


,순위,상품코드,판매상품 ID,상품명,정가,판매가,할인율,적립율,적립예정포인트,인물,출판사,발행(출시)일자,분야
0,1,9788936434120,S000000610612,소년이 온다,"15,000","13,500",10%,5%,750,한강,창비,20140519,소설
1,2,9788998441012,S000001632467,모순,"13,000","11,700",10%,5%,650,양귀자,쓰다,20130401,소설
2,3,9788997780624,S000216262757,결국 국민이 합니다,"22,000","19,800",10%,5%,"1,100",이재명,오마이북,20250415,정치/사회
3,4,9788936439743,S000216111714,혼모노,"18,000","16,200",10%,5%,900,성해나,창비,20250328,소설
4,5,9788937473401,S000200550189,급류,"14,000","12,600",10%,5%,700,정대건,민음사,20221222,소설


In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 199 entries, 0 to 198
Data columns (total 13 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   순위        199 non-null    int64
 1   상품코드      199 non-null    int64
 2   판매상품 ID   199 non-null    str  
 3   상품명       199 non-null    str  
 4   정가        199 non-null    str  
 5   판매가       199 non-null    str  
 6   할인율       199 non-null    str  
 7   적립율       199 non-null    str  
 8   적립예정포인트   199 non-null    str  
 9   인물        199 non-null    str  
 10  출판사       199 non-null    str  
 11  발행(출시)일자  199 non-null    int64
 12  분야        199 non-null    str  
dtypes: int64(3), str(10)
memory usage: 20.3 KB


In [5]:
df.isna().sum().sum()

np.int64(0)

### 실행 결과 요약

- 데이터는 **199행, 13개 컬럼**으로 구성되어 있다.
- 주요 컬럼으로 `순위`, `상품코드`, `상품명`, `정가`, `판매가`, `인물`, `출판사`, `분야` 등이 있다.
- `순위`, `상품코드`, `발행(출시)일자`는 정수형(`int64`)이고, 나머지 대부분은 문자열(`str`)로 확인되었다.
- 모든 컬럼의 **결측치가 0개**로 확인되어 비어 있는 값은 없다.
- `head()`를 통해 앞의 5행을 확인하여 데이터가 정상적으로 불러와졌음을 확인하였다.

### 실습 3. 분석에 필요한 컬럼만 선택하기

원본 데이터의 모든 컬럼이 이후 분석에 필요한 것은 아닙니다.

이번 과정에서는 다음 컬럼을 먼저 선택합니다.

In [6]:
# 사용할 컬럼만 선택해서 새로운 DataFrame 만들기
df_books = df[[
    "순위",
    "판매상품 ID",
    "상품명",
    "판매가",
    "인물",
    "출판사",
    "발행(출시)일자",
    "분야"
]]

# 선택한 컬럼 확인
print(df_books.columns)

# 데이터 앞 5행 확인
display(df_books.head())

Index(['순위', '판매상품 ID', '상품명', '판매가', '인물', '출판사', '발행(출시)일자', '분야'], dtype='str')


,순위,판매상품 ID,상품명,판매가,인물,출판사,발행(출시)일자,분야
0,1,S000000610612,소년이 온다,"13,500",한강,창비,20140519,소설
1,2,S000001632467,모순,"11,700",양귀자,쓰다,20130401,소설
2,3,S000216262757,결국 국민이 합니다,"19,800",이재명,오마이북,20250415,정치/사회
3,4,S000216111714,혼모노,"16,200",성해나,창비,20250328,소설
4,5,S000200550189,급류,"12,600",정대건,민음사,20221222,소설


### 실행 결과 요약

- 원본 데이터에서 분석에 필요한 **8개 컬럼만 선택**하여 `df_books`를 만들었다.
- 선택한 컬럼은 `순위`, `판매상품 ID`, `상품명`, `판매가`, `인물`, `출판사`, `발행(출시)일자`, `분야`이다.
- `df_books.columns`와 `df_books.head()`를 통해 원하는 컬럼과 데이터가 정상적으로 선택된 것을 확인하였다.

### 실습 4. 컬럼 이름을 사용하기 쉽게 정리하기

원본 컬럼 이름은 사람이 읽기에는 좋지만 코드에서 반복해서 사용하기에는 조금 길 수 있습니다.

이번 실습에서는 다음처럼 간단하게 정리할 수 있습니다.


In [7]:
# 컬럼 이름 변경
df_books = df_books.rename(columns={
    "판매상품 ID": "판매상품ID",
    "인물": "저자",
    "발행(출시)일자": "발행일"
})

# 변경된 컬럼 이름 확인
print(df_books.columns)

Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')


### 실행 결과 요약

- `rename()`을 사용하여 일부 컬럼 이름을 더 간단하게 변경하였다.
- `판매상품 ID` → `판매상품ID`, `인물` → `저자`, `발행(출시)일자` → `발행일`로 변경하였다.
- `df_books.columns`를 통해 컬럼 이름이 정상적으로 변경된 것을 확인하였다.

### 실습 5. 결측치 확인하기

결측치는 값이 비어 있는 데이터를 의미합니다.

데이터 분석에서는 결측치를 무조건 삭제하지 않습니다.

먼저 어디에 얼마나 있는지 확인하고, 분석 목적에 맞게 처리 방법을 결정합니다.


In [8]:
# 각 컬럼의 결측치 개수 확인
print(df_books.isna().sum())
# 결측치 개수 계산
missing = df_books.isna().sum()

# 결측치가 1개 이상 있는 컬럼만 확인
print(missing[missing > 0])

순위        0
판매상품ID    0
상품명       0
판매가       0
저자        0
출판사       0
발행일       0
분야        0
dtype: int64
Series([], dtype: int64)


### 실행 결과 요약

- `isna().sum()`을 사용하여 각 컬럼의 결측치 개수를 확인하였다.
- 모든 컬럼의 결측치가 **0개**로 확인되었다.
- 따라서 현재 `df_books`에는 결측치가 없으며, 결측치에 대한 별도의 처리는 필요하지 않다.

### 실습 6. 판매가를 숫자로 바꾸기

Excel에서 판매가가 다음과 같이 저장되어 있을 수 있습니다.


In [9]:
# 1. 판매가에서 쉼표(,) 제거
df_books["판매가"] = df_books["판매가"].str.replace(",", "")

# 2. 판매가를 숫자형으로 변환
df_books["판매가"] = pd.to_numeric(df_books["판매가"])

# 3. 변환된 자료형 확인
print(df_books["판매가"].dtype)

# 데이터 확인
display(df_books.head())

int64


,순위,판매상품ID,상품명,판매가,저자,출판사,발행일,분야
0,1,S000000610612,소년이 온다,13500,한강,창비,20140519,소설
1,2,S000001632467,모순,11700,양귀자,쓰다,20130401,소설
2,3,S000216262757,결국 국민이 합니다,19800,이재명,오마이북,20250415,정치/사회
3,4,S000216111714,혼모노,16200,성해나,창비,20250328,소설
4,5,S000200550189,급류,12600,정대건,민음사,20221222,소설


### 실행 결과 요약

- `판매가`에 포함된 쉼표(`,`)를 제거하였다.
- `pd.to_numeric()`을 사용하여 `판매가`를 문자열에서 숫자형으로 변환하였다.
- 변환 후 `판매가`의 자료형이 `int64`인 것을 확인하였다.
- 따라서 판매가를 이용한 평균, 합계 등의 숫자 계산이 가능해졌다.

### 실습 7. 발행일을 날짜형으로 바꾸기

원본 데이터의 발행일은 다음처럼 보일 수 있습니다.

In [10]:
# 발행일을 날짜 자료형으로 변환
df_books["발행일"] = pd.to_datetime(
    df_books["발행일"],
    format="%Y%m%d"
)

# 변환된 자료형 확인
print(df_books["발행일"].dtype)

# 앞의 5개 값 확인
print(df_books["발행일"].head())

datetime64[us]
0   2014-05-19
1   2013-04-01
2   2025-04-15
3   2025-03-28
4   2022-12-22
Name: 발행일, dtype: datetime64[us]


### 실행 결과 요약

- `pd.to_datetime()`을 사용하여 `발행일`을 날짜 자료형으로 변환하였다.
- `YYYYMMDD` 형식의 값이 `YYYY-MM-DD` 형태의 날짜로 정상적으로 변환되었다.
- 변환 후 `발행일`의 자료형이 `datetime64[us]`로 확인되어 날짜 데이터로 사용할 수 있게 되었다.

### 실습 8. 문자열 앞뒤 공백 정리하기

상품명, 저자, 출판사, 분야는 문자열 데이터입니다.

문자열 앞뒤에 불필요한 공백이 있으면 같은 값도 서로 다른 데이터처럼 처리될 수 있습니다.

예를 들어 다음 두 값은 사람이 보기에는 같지만 문자열로는 다릅니다.

In [11]:
# 공백을 제거할 문자열 컬럼 목록
text_columns = ["상품명", "저자", "출판사", "분야"]

# 각 컬럼의 앞뒤 공백 제거
for col in text_columns:
    df_books[col] = df_books[col].str.strip()

# 처리 결과 확인
print(df_books[text_columns].head())

          상품명   저자   출판사     분야
0      소년이 온다   한강    창비     소설
1          모순  양귀자    쓰다     소설
2  결국 국민이 합니다  이재명  오마이북  정치/사회
3         혼모노  성해나    창비     소설
4          급류  정대건   민음사     소설


### 실행 결과 요약

- `상품명`, `저자`, `출판사`, `분야` 컬럼의 문자열 앞뒤 공백을 제거하였다.
- 반복문과 `str.strip()`을 사용하여 여러 문자열 컬럼을 한 번에 정리하였다.
- `head()`를 통해 정리된 데이터의 앞 5행을 확인하였다.
- 불필요한 공백을 제거하여 이후 문자열 비교나 분석을 더 정확하게 할 수 있도록 정리하였다.

### 실습 9. 중복 데이터 확인하기

데이터에서 중복을 발견했다고 해서 바로 삭제하면 안 됩니다.

먼저 무엇을 기준으로 중복인지 생각해야 합니다.

예를 들어 같은 상품명이 여러 번 등장할 수 있습니다.

하지만 같은 제목이라도 다른 판본, 다른 상품, 개정판 등이 있을 수 있습니다.

따라서 이번 데이터에서는 상품명만 보고 중복 행을 바로 삭제하지 않습니다.

우선 상품을 구분하는 판매상품ID를 확인합니다.

In [12]:
# 1. 전체 행이 완전히 같은 중복 행 개수
print("전체 중복 행:", df_books.duplicated().sum())

# 2. 판매상품ID가 중복된 행 개수
print("판매상품ID 중복:", df_books.duplicated(subset=["판매상품ID"]).sum())

# 3. 상품명이 중복된 행 개수
print("상품명 중복:", df_books.duplicated(subset=["상품명"]).sum())

전체 중복 행: 0
판매상품ID 중복: 0
상품명 중복: 0


### 실행 결과 요약

- 전체 행이 완전히 같은 중복 데이터는 **0개**로 확인되었다.
- `판매상품ID`를 기준으로 한 중복도 **0개**였다.
- `상품명`을 기준으로 한 중복도 **0개**였다.
- 따라서 현재 `df_books`에서는 확인한 기준에 따른 중복 데이터가 없으며, 별도의 중복 제거가 필요하지 않다.

### 실습 10. 전처리 결과 검증하기

지금까지 여러 전처리를 수행했습니다.
마지막에는 반드시 데이터가 의도한 형태가 되었는지 다시 확인합니다.

In [13]:
# 1. 전체 행과 컬럼 개수
print("1. 데이터 크기")
print(df_books.shape)

# 2. 컬럼 이름
print("\n2. 컬럼 이름")
print(df_books.columns)

# 3. 각 컬럼의 자료형
print("\n3. 자료형")
print(df_books.dtypes)

# 4. 각 컬럼의 결측치 개수
print("\n4. 결측치 개수")
print(df_books.isna().sum())

# 5. 판매가 앞의 5개 값
print("\n5. 판매가 앞의 5개 값")
print(df_books["판매가"].head())

# 6. 발행일 앞의 5개 값
print("\n6. 발행일 앞의 5개 값")
print(df_books["발행일"].head())

# 7. 판매상품ID 중복 개수
print("\n7. 판매상품ID 중복 개수")
print(df_books.duplicated(subset=["판매상품ID"]).sum())

# 8. 전체 데이터 앞의 5행
print("\n8. 데이터 앞의 5행")
display(df_books.head())

1. 데이터 크기
(199, 8)

2. 컬럼 이름
Index(['순위', '판매상품ID', '상품명', '판매가', '저자', '출판사', '발행일', '분야'], dtype='str')

3. 자료형
순위                 int64
판매상품ID               str
상품명                  str
판매가                int64
저자                   str
출판사                  str
발행일       datetime64[us]
분야                   str
dtype: object

4. 결측치 개수
순위        0
판매상품ID    0
상품명       0
판매가       0
저자        0
출판사       0
발행일       0
분야        0
dtype: int64

5. 판매가 앞의 5개 값
0    13500
1    11700
2    19800
3    16200
4    12600
Name: 판매가, dtype: int64

6. 발행일 앞의 5개 값
0   2014-05-19
1   2013-04-01
2   2025-04-15
3   2025-03-28
4   2022-12-22
Name: 발행일, dtype: datetime64[us]

7. 판매상품ID 중복 개수
0

8. 데이터 앞의 5행


,순위,판매상품ID,상품명,판매가,저자,출판사,발행일,분야
0,1,S000000610612,소년이 온다,13500,한강,창비,2014-05-19,소설
1,2,S000001632467,모순,11700,양귀자,쓰다,2013-04-01,소설
2,3,S000216262757,결국 국민이 합니다,19800,이재명,오마이북,2025-04-15,정치/사회
3,4,S000216111714,혼모노,16200,성해나,창비,2025-03-28,소설
4,5,S000200550189,급류,12600,정대건,민음사,2022-12-22,소설


### 전처리 결과 요약

- 최종 데이터는 **199행, 8개 컬럼**으로 구성되어 있다.
- `판매가`는 `int64`로 변환되어 숫자 계산이 가능한 형태이다.
- `발행일`은 `datetime64[us]`로 변환되어 날짜 데이터로 사용할 수 있다.
- 모든 컬럼의 **결측치는 0개**이다.
- `판매상품ID`를 기준으로 한 **중복 데이터는 0개**이다.
- 최종 데이터의 앞 5행을 확인하여 전처리 결과가 정상적으로 반영된 것을 확인하였다.

### 실습 11. 전처리 데이터 저장하기

다음 Chapter에서는 이 데이터를 다시 사용합니다.

따라서 전처리한 DataFrame을 CSV 파일로 저장합니다.

추천 파일명은 다음과 같습니다.

In [14]:
df_books.to_csv("book_bestseller_clean.csv", encoding="utf-8-sig", index=False)